app/services/course/material/zip_download_service.rb

```diff
@@ -1,12 +1,70 @@
 class Course::Material::ZipDownloadService
-  class << self
-    # Downloads the materials and zip them.
-    #
-    # @param [Course::Material::Folder] folder The folder containing the materials.
-    # @param [Array<Course::Material>] materials The materials to be downloaded.
-    # @return [String] The path to the zip file.
-    def download_and_zip(_folder, _materials)
-      # TODO: To implement
+  attr_accessor :materials, :folder, :base_dir
+
+  # @param [Course::Material::Folder] folder The folder containing the materials.
+  # @param [Array<Course::Material>] materials The materials to be downloaded.
+  def initialize(folder, materials)
+    @folder = folder
+    @materials = Array(materials)
+    @base_dir = Dir.mktmpdir('coursemology-download-')
+  end
+
+  # Downloads the materials and zip them.
+  #
+  # @return [String] The path to the zip file.
+  def download_and_zip
+    download_to(base_dir)
+    zip_dir(base_dir)
+  end
+
+  private
+
+  # Downloads the materials to the the given directory.
+  #
+  # @param [String] dir The path to the directory.
+  def download_to(dir)
+    materials.each do |material|
+      download_material(material, folder, dir)
+    end
+  end
+
+  # Zip the directory and write to the file.
+  # @param [String] dir The path to the directory.
+  #
+  # @return [String] The path to the zip file.
+  def zip_dir(dir)
+    output_file = dir + '.zip'
+    Zip::File.open(output_file, Zip::File::CREATE) do |zip_file|
+      Dir["#{dir}/**/**"].each do |file|
+        zip_file.add(file.sub(File.join(dir + '/'), ''), file)
+      end
     end
+
+    output_file
+  end
+
+  # Downloads the material and store it in the given directory.
+  def download_material(material, folder, dir)
+    path_of_folder = folder_path(folder)
+    path_of_material = material_path(material)
+    file_path = Pathname.new(dir) + path_of_material.relative_path_from(path_of_folder)
+
+    file_path.dirname.mkpath
+    file = File.open(file_path, 'wb')
+    file << material.attachment.file_upload.read
+    file.close
+  end
+
+  # Returns the path of the folder, note that '/' will be returned for root_folder
+  def folder_path(folder)
+    folders = folder.ancestors.reverse + [folder]
+    folders.shift # Remove the root folder
+    path = File.join('/', folders.map { |f| Pathname.normalize_filename(f.name) })
```

Comment:
After our discuss, I lean to use option 1 now, and handle all these in models.  Think models need to make sure that everything is correct.